treeherder/perf/investigation_aids.py

```diff
@@ -0,0 +1,216 @@
+                                  cls._task_in_context(action['context'], original_task['tags']))
+
+            if (action_of_known_kind and action_is_relevant and
+                    (no_context_or_task_to_check or task_is_in_context)):
+                accumulating_actions.append(action)
+            return accumulating_actions
+
+        actions = reduce(collect_relevant_actions, actions, [])
+        return list(actions)
+
+    @classmethod
+    def _has_known_kind(cls, action):
+        return action['kind'] in cls.KNOWN_KINDS
+
+    @classmethod
+    def _task_in_context(cls, tag_set_list, task_tags):
+        return any(
+            all(task_tags.get(tag) and task_tags[tag] == tag_set[tag]
+                for tag in tag_set.keys())
+            for tag_set in tag_set_list
+        )
+
+
+class BackfillTool:
+
+    def __init__(self, taskcluster_model: TaskclusterModel):
+        self.tc_model = taskcluster_model
+
+    def backfill_job(self, job_id: str) -> str:
+        job = self._fetch_job(job_id)
+
+        if not self.can_backfill(job):
+            logger.warning(f"Cannot backfill job {job_id}.")
+            return ''
+
+        decision_job = self._fetch_associated_decision_job(job)
+        decision_task_id = decision_job.taskcluster_metadata.task_id
+
+        results = self.tc_model.load(decision_task_id, job)
+        backfill_task = self._get_action(results["actions"], "backfill")
+
+        logger.debug(f"Requesting backfill for task {results['originalTaskId']}...")
+        task_id = self.tc_model.submit(
+            action=backfill_task,
+            decision_task_id=decision_task_id,
+            task_id=results["originalTaskId"],
+            input={},
+            static_action_variables=results["staticActionVariables"]
+        )
+        return task_id
+
+    def can_backfill(self, job: Job) -> bool:
+        has_credentials = self.tc_model.has_credentials()
+        is_try_repo = job.repository.is_try_repo
+
+        if not has_credentials:
+            logger.debug("Credentials for backfilling are missing.")
+        if is_try_repo:
+            logger.debug("Try repository isn't suited for backfilling.")
+        return has_credentials and not is_try_repo
+
+    @staticmethod
+    def _fetch_job(job_id: str) -> Job:
+        try:
+            return Job.objects.get(id=job_id)
+        except ObjectDoesNotExist:
+            raise LookupError(f"Job {job_id} not found.")
+
+    @staticmethod
+    def _fetch_associated_decision_job(job) -> Job:
+        logger.debug(f"Fetching decision task of job {job.id}...")
+        decision_type = JobType.objects.filter(name="Gecko Decision Task",
+                                               symbol="D")
+        return Job.objects.get(repository_id=job.repository_id,
+                               job_type_id=Subquery(decision_type.values('id')[:1]),
+                               push_id=job.push_id)
```

Comment:
I'm not familiar enough with treeherder to really comment on these, but I suspect these might belong in `treeherder.model.models` rather than here?

Reply:
Yes, good point.